### PR TITLE
Using events to signal the enabled Security Group for Pods feature

### PR DESCRIPTION
**Pull request by haouc**
**What type of PR is this?**
The PR is changing how CNI send signals to [amazon-vpc-resource-controller](https://github.com/aws/amazon-vpc-resource-controller-k8s/pull/157) for enabled feature of Security Group for Pods. Amazon VPC CNI is labeling nodes object to send the signal to the controller when the feature is enabled by user. We should remove the permission labeling node and use alternatives to handle the process.

feature
**Which issue does this PR fix**:
N/A

**What does this PR do / Why do we need it**:
This PR changes the way sending signal to another Amazon service.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:
N/A

**Testing done on this change**:

Tested in EKS clusters to test if the nodes can be enabled and pods using Security Group can be created successfully.
```
% k get events | grep AwsNodeNotificationToRc
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal    vpc.amazonaws.com/eniConfig=us-west-2a
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal    vpc.amazonaws.com/has-trunk-attached=false
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal    vpc.amazonaws.com/has-trunk-attached=false
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal   vpc.amazonaws.com/eniConfig=us-west-2b
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal   vpc.amazonaws.com/has-trunk-attached=false
18m         Normal    AwsNodeNotificationToRc   node/ip-192-168-x-x.us-west-2.compute.internal   vpc.amazonaws.com/has-trunk-attached=false
```
**Automation added to e2e**:

This PR doesn't change functionality and no need to add new e2e modules.
**Will this PR introduce any new dependencies?**:

No
**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
Yes. This has to be used with new version of amazon-vpc-resource-controller. 

**Does this change require updates to the CNI daemonset config files to work?**:

not necessary. but if downgrade VPC CNI without adding back node update permission, security group for pods will not work with older version of VPC CNI.
**Does this PR introduce any user-facing change?**:

Yes, nodes will no longer be modified by VPC CNI and new node events will be emitted.
```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/awsutils/awsutils_test.go [hunk 1]
```diff
@@ -139,15 +169,15 @@ func TestInitWithEC2metadata(t *testing.T) {
...
-		assert.Equal(t, ins.instanceID, instanceID)
-		assert.Equal(t, ins.primaryENImac, primaryMAC)
-		assert.Equal(t, ins.primaryENI, primaryeniID)
-		assert.Equal(t, subnetID, ins.subnetID)
+		assert.Equal(t, az, cache.availabilityZone)
+		assert.Equal(t, localIP, cache.localIPv4.String())
+		assert.Equal(t, cache.instanceID, instanceID)
+		assert.Equal(t, cache.primaryENImac, primaryMAC)
+		assert.Equal(t, cache.primaryENI, primaryeniID)
+		assert.Equal(t, subnetID, cache.subnetID)
```
Comment (jdn5126) [inline pkg/awsutils/awsutils_test.go:180]: nit: should we check the value of `cache.eventRecorder` here?

File: pkg/awsutils/awsutils_test.go [hunk 2]
```diff
@@ -380,11 +410,12 @@ func TestAllocENI(t *testing.T) {
 	mockEC2.EXPECT().AttachNetworkInterfaceWithContext(gomock.Any(), gomock.Any(), gomock.Any()).Return(attachResult, nil)
 	mockEC2.EXPECT().ModifyNetworkInterfaceAttributeWithContext(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, nil)
 
-	ins := &EC2InstanceMetadataCache{
+	cache := &EC2InstanceMetadataCache{
 		ec2SVC: mockEC2,
 		imds:   TypedIMDS{mockMetadata},
 	}
-	_, err := ins.AllocENI(false, nil, "")
+	cache.eventRecorder = setupEventRecorder(t)
```
Comment (jdn5126) [inline pkg/awsutils/awsutils_test.go:417]: nit: can we set `eventRecorder` in constructor above? Seems easier to read that way. Same with the ones below

File: pkg/ipamd/ipamd.go [hunk 3]
```diff
@@ -374,9 +376,15 @@ func New(rawK8SClient client.Client, cachedK8SClient client.Client) (*IPAMContex
...
+	eventrecorder, err := eventrecorder.New(rawK8SClient, cachedK8SClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "ipamd: can not initialize event recorder")
+	}
+	c.eventRecorder = eventrecorder
+
 	c.disableENIProvisioning = disablingENIProvisioning()
 
-	client, err := awsutils.New(c.useCustomNetworking, c.disableENIProvisioning, c.enableIPv4, c.enableIPv6)
+	client, err := awsutils.New(c.useCustomNetworking, c.disableENIProvisioning, c.enableIPv4, c.enableIPv6, eventrecorder)
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:387]: nit: `c.eventRecorder`

File: pkg/ipamd/ipamd.go [hunk 4]
```diff
@@ -1319,13 +1311,8 @@ func (c *IPAMContext) nodeIPPoolReconcile(ctx context.Context, interval time.Dur
...
 
 		if c.enablePodENI && metadataResult.TrunkENI != "" {
-			// Label the node that we have a trunk
-			err = c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-			if err != nil {
-				podENIErrInc("askForTrunkENIIfNeeded")
-				log.Errorf("Failed to set node label for trunk. Aborting reconcile", err)
-				return
-			}
+			// Signal event to VPC RC that the node has a trunk already
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:1314]: nit: we are not signaling anything to VPC RC here, right? Just logging locally as the annotation is already present

File: pkg/ipamd/ipamd.go [hunk 5]
```diff
@@ -1936,7 +1880,7 @@ func (c *IPAMContext) GetPod(podName, namespace string) (*corev1.Pod, error) {
 	}
 	err := c.rawK8SClient.Get(ctx, podKey, &pod)
 	if err != nil {
-		return nil, fmt.Errorf("Error while trying to retrieve Pod Info: %s", err)
+		return nil, fmt.Errorf("Error while trying to retrieve Pod Info: %s", err.Error())
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:1883]: nit: `%v", err`

File: pkg/ipamd/ipamd_test.go [hunk 6]
```diff
@@ -1850,6 +1850,26 @@ func TestIPAMContext_askForTrunkENIIfNeeded(t *testing.T) {
...
+	fakeRecorder := events.NewFakeRecorder(3)
+	k8sSchema := runtime.NewScheme()
+	clientgoscheme.AddToScheme(k8sSchema)
+
+	mockEventRecorder := &eventrecorder.EventRecorder{
+		Recorder:        fakeRecorder,
+		RawK8SClient:    testclient.NewClientBuilder().WithScheme(k8sSchema).Build(),
+		CachedK8SClient: testclient.NewClientBuilder().WithScheme(k8sSchema).Build(),
+	}
+	eventrecorder.MyNodeName = myNodeName
```
Comment (jdn5126) [inline pkg/ipamd/ipamd_test.go:1862]: nit: this is strange to read, can you add a comment that this is setting global var in `eventrecorder` package

File: test/framework/framework.go [hunk 7]
```diff
@@ -53,9 +60,28 @@ func New(options Options) *Framework {
 	eniConfig.AddToScheme(k8sSchema)
 	sgp.AddToScheme(k8sSchema)
 
-	k8sClient, err := client.New(config, client.Options{Scheme: k8sSchema})
+	stopChan := ctrl.SetupSignalHandler()
```
Comment (jdn5126) [inline test/framework/framework.go:63]: nit: can you add some comments here? A but hard to follow what is happening

**Review by jdn5126 — COMMENTED** [review]
Changes overall LGTM, just some nits and I will give @jayanthvn a chance to review

File: pkg/ipamd/ipamd.go [hunk 8]
```diff
@@ -1201,15 +1196,12 @@ func (c *IPAMContext) askForTrunkENIIfNeeded(ctx context.Context) {
...
 			return
 		}
 		// We need to signal that VPC Resource Controller needs to attach a trunk ENI
-		err := c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "false")
-		if err != nil {
-			podENIErrInc("askForTrunkENIIfNeeded")
-			log.Errorf("Failed to set node label", err)
-		}
+		log.Debug("VPC CNI is asking RC to initialize trunk interface")
+		c.eventRecorder.SendNodeEvent(corev1.EventTypeNormal, eventrecorder.EventReason, "NeedTrunk", "vpc.amazonaws.com/has-trunk-attached=false")
```
Comment (jdn5126) [inline pkg/ipamd/ipamd.go:1204]: If node does not support `has-trunk-attached`, can we avoid sending this event and log error for readability?

File: pkg/ipamd/ipamd.go (hunk 4 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1314]: Right, this was leftover after I removed the call to send event during updating some logic flow. Thanks for catching this.

File: pkg/ipamd/ipamd.go (hunk 5 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1883]: Why should we log the entire object instead of just error string for information?

File: pkg/ipamd/ipamd.go (hunk 8 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1204]: It is a bit harder for now because we haven't verified if we can validate node type for the feature during runtime. But this is planned to be done for the next release of VPC resource controller if APIs are available from EC2. You can refer to t[his open issue](https://github.com/aws/amazon-vpc-resource-controller-k8s/issues/98).
Otherwise we have to maintain a hard copy of file listing all supported types, which is exactly what we want to deprecate.

File: pkg/ipamd/ipamd.go [hunk 9]
```diff
@@ -538,37 +550,20 @@ func (c *IPAMContext) nodeInit() error {
...
-	if metadataResult.TrunkENI != "" {
-		// Signal to VPC Resource Controller that the node has a trunk already
-		err := c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-		if err != nil {
-			log.Errorf("Failed to set node label", err)
-			podENIErrInc("nodeInit")
-			// If this fails, we probably can't talk to the API server. Let the pod restart
-			return err
+		if metadataResult.TrunkENI != "" {
+			// Signal event to VPC RC that the node has a trunk already
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:561]: Don't we want to signal the event to VPC RC about `trunk attached = true` or since it is found in metadataResult assumption is the node is already managed in VPC RC? If it is not needed let's remove the comment.

Comment (haouc) [reply pkg/ipamd/ipamd.go:561]: Right, I think we don't have to send a confirmation back to VPC CNI since CNI has it' own check on trunk interface. I will remove the meaningless comment here after all comments are collect for the current revision. Thanks @jayanthvn

File: pkg/awsutils/awsutils.go [hunk 10]
```diff
@@ -433,7 +436,6 @@ func New(useCustomNetworking, disableENIProvisioning, v4Enabled, v6Enabled bool)
 		return nil, err
 	}
 	// event recorder to raise events for failed EC2 API calls
```
Comment (jdn5126) [inline pkg/awsutils/awsutils.go:438]: nit: comment can be deleted

**Review by jdn5126 — COMMENTED** [review]
Changes LGTM, will give Jayanth a chance to finish review

File: pkg/ipamd/ipamd.go (hunk 5 above)
Comment (jdn5126) [reply pkg/ipamd/ipamd.go]: Good point, this is fine

File: pkg/eniconfig/eniconfig.go [hunk 11]
```diff
@@ -141,10 +142,9 @@ func GetNodeSpecificENIConfigName(ctx context.Context, k8sClient client.Client)
 	}
 
 	eniConfigName = val
-	if val != eniConfigDefault {
-		labels := node.GetLabels()
-		labels["vpc.amazonaws.com/eniConfig"] = eniConfigName
-		node.SetLabels(labels)
+	if val != eniConfigDefault && generate_event {
+		// Signal event to VPC RC that the custom networking is enabled
+			eventRecorder.SendNodeEvent(corev1.EventTypeNormal, "AwsNodeNotification", "CustomNetworkingEnabled", eniConfigLabel+"="+eniConfigName)
```
Comment (achevuru) [inline pkg/eniconfig/eniconfig.go]: Why generate the event from here though? Why not from the IPAMD init flow where we currently check the env variable and are currently labeling the node? - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/ipamd/ipamd.go#L543. We can avoid the eventRecorder references in the controller code flow here...

File: pkg/ipamd/ipamd.go [hunk 12]
```diff
@@ -374,9 +376,15 @@ func New(rawK8SClient client.Client, cachedK8SClient client.Client) (*IPAMContex
 	c.enableIPv4 = isIPv4Enabled()
 	c.enableIPv6 = isIPv6Enabled()
 
+	eventRecorder, err := eventrecorder.New(rawK8SClient, cachedK8SClient)
+ 	if err != nil {
+ 		return nil, errors.Wrap(err, "ipamd: can not initialize event recorder")
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go]: nit: `can not` --> `unable to`. `can not` sounds like a permission related error to me.

File: pkg/ipamd/ipamd.go [hunk 13]
```diff
@@ -538,37 +547,22 @@ func (c *IPAMContext) nodeInit() error {
...
-		if err != nil {
-			log.Errorf("Failed to delete eniConfig node label", err)
-			podENIErrInc("nodeInit")
-			return err
+	if c.enablePodENI {
+		if c.useCustomNetworking {
+			eniConfigName, err := eniconfig.GetNodeSpecificENIConfigName(ctx, c.cachedK8SClient, c.eventRecorder, true)
+			if err == nil && eniConfigName != "default" {
+				// Signal event to VPC RC that the custom networking is enabled
+				c.eventRecorder.SendNodeEvent(corev1.EventTypeNormal, "AwsNodeNotification", "CustomNetworkingEnabled", vpcENIConfigLabel+"="+eniConfigName)
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go]: I see that we are generating the event for Custom networking from here as well...Why do we need to do it again from `GetNodeSpecificENIConfigName`?

File: pkg/ipamd/ipamd.go (hunk 9 above)
Comment (achevuru) [reply pkg/ipamd/ipamd.go:561]: But, this is done only once on bootup isn't it? Also, how will VPC RC rebuild the cache upon a restart?

File: pkg/ipamd/ipamd.go [hunk 14]
```diff
@@ -834,7 +829,7 @@ func (c *IPAMContext) tryAllocateENI(ctx context.Context) error {
 	var subnet string
 
 	if c.useCustomNetworking {
-		eniCfg, err := eniconfig.MyENIConfig(ctx, c.cachedK8SClient)
+		eniCfg, err := eniconfig.MyENIConfig(ctx, c.cachedK8SClient, c.eventRecorder, c.enablePodENI)
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:833]: Same comment as above. Do we need to do this?

File: pkg/ipamd/ipamd.go (hunk 9 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:561]: Do you mean VPC CNI check trunk? In current code, VPC CNI does reconcile this every 30s in update IP resources and rely on datastore for trunk info. Upon restart of VPC RC, it iterates through all nodes and manages all nodes which have valid labels. Their trunk interfaces are also described and cached.

Comment (achevuru) [reply pkg/ipamd/ipamd.go:561]: Right now, yes. But once we transition to events based approach, VPC RC would've to go through the events against the node to do the same, isn't it?

Comment (haouc) [reply pkg/ipamd/ipamd.go:561]: Once we change to event, RC relies on event and node label. The workflow looks like this

1. CNI emits events continuously till finding trunk interface on node 
2. RC reconcile events and label node if event is valid
3. RC provision interfaces based on node label
4. if RC restarts, RC looks for node label for building cache

If events are still alive during RC booting up, RC will 

1. if the node is in cache, RC will skip the node
2. if the node hasn't been added in cache, RC will retry to label the node which should be safe

File: pkg/ipamd/ipamd.go (hunk 13 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go]: This is used to tell VPC RC to look for specific subnet when custom networking is enabled. CNI does this work for RC before events.

File: pkg/eniconfig/eniconfig.go (hunk 11 above)
Comment (haouc) [reply pkg/eniconfig/eniconfig.go]: Thanks, Apurup. I have move the code to ipamd instead.

File: pkg/ipamd/ipamd.go [hunk 15]
```diff
@@ -1319,13 +1365,7 @@ func (c *IPAMContext) nodeIPPoolReconcile(ctx context.Context, interval time.Dur
...
 
 		if c.enablePodENI && metadataResult.TrunkENI != "" {
-			// Label the node that we have a trunk
-			err = c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-			if err != nil {
-				podENIErrInc("askForTrunkENIIfNeeded")
-				log.Errorf("Failed to set node label for trunk. Aborting reconcile", err)
-				return
-			}
+			log.Debugf("Trunk interface (%s) has been added to the node already.", metadataResult.TrunkENI)
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:1368]: Are we making the assumption here that the event would've been already generated for this node, since the trunk is already present?

File: pkg/ipamd/ipamd.go [hunk 16]
```diff
@@ -1947,7 +1944,7 @@ func (c *IPAMContext) AnnotatePod(podName, podNamespace, key, val string) error
 	var err error
 
 	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-		pod := &corev1.Pod{}
+		var pod *corev1.Pod
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:1947]: Why the change here?

File: pkg/sgpp/constants.go [hunk 17]
```diff
@@ -3,8 +3,12 @@ package sgpp
 type EnforcingMode string
 
 const (
-	EnforcingModeStrict   EnforcingMode = "strict"
-	EnforcingModeStandard EnforcingMode = "standard"
+	EnforcingModeStrict               EnforcingMode = "strict"
+	EnforcingModeStandard             EnforcingMode = "standard"
+	VpcCNINodeEventActionForTrunk     string        = "NeedTrunk"
+	TrunkEventNote                    string        = "vpc.amazonaws.com/has-trunk-attached=false"
```
Comment (achevuru) [inline pkg/sgpp/constants.go:9]: Are we not generating event for `vpc.amazonaws.com/has-trunk-attached=true`? I believe we were trying to address that for nodes that come up with a trunk interface attached.

File: pkg/utils/eventrecorder/eventrecorder.go [hunk 18]
```diff
@@ -16,84 +16,107 @@ package eventrecorder
...
-		e.recorder.Event(&pod, eventType, reason, message)
+		e.Recorder.Eventf(&pod, nil, eventType, reason, "", message)
 	}
 }
+
+func (e *EventRecorder) findMyNode(ctx context.Context) (corev1.Node, error) {
+	var node corev1.Node
+	// Find my node
+	err := e.CachedK8SClient.Get(ctx, types.NamespacedName{Name: MyNodeName}, &node)
+	log.Debugf("Node found %s - labels - %d", node.Name, len(node.Labels))
```
Comment (achevuru) [inline pkg/utils/eventrecorder/eventrecorder.go:98]: We're not checking for error here before we print this debug message. Will be misleading.

File: pkg/ipamd/ipamd.go [hunk 19]
```diff
@@ -538,36 +553,12 @@ func (c *IPAMContext) nodeInit() error {
...
-		err := c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-		if err != nil {
-			log.Errorf("Failed to set node label", err)
-			podENIErrInc("nodeInit")
-			// If this fails, we probably can't talk to the API server. Let the pod restart
-			return err
-		}
-	} else {
-		// Check if we want to ask for one
+		// we want to ask VPC RC once during node init to cover special cases
```
[the remark this answers is not in the document]
Comment (haouc) [reply pkg/ipamd/ipamd.go:561]: The error comes from sending node events. If error happens, we don't have to retry since the CNI will resend another event in 30 s. We will just let it fail through.

File: pkg/ipamd/ipamd.go [hunk 20]
```diff
@@ -655,7 +646,15 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
 }
 
 func (c *IPAMContext) updateIPPoolIfRequired(ctx context.Context) {
-	c.askForTrunkENIIfNeeded(ctx)
+	if c.enablePodENI {
+		c.askForTrunkENIIfNeeded(ctx)
```
[the remark this answers is not in the document]
Comment (haouc) [reply pkg/ipamd/ipamd.go:650]: For the same reason, we let CNI tries to send another event in 30s instead of retry or handle error at this place since the error should come from `SendNodeEvent()`.

File: pkg/ipamd/ipamd.go [hunk 21]
```diff
@@ -1197,20 +1197,66 @@ func (c *IPAMContext) logPoolStats(dataStoreStats *datastore.DataStoreStats) {
...
+	}
+
+	return err
+}
+
+func (c *IPAMContext) getNodeLabels(ctx context.Context) (map[string]string, error) {
+	var node corev1.Node
+	err := c.cachedK8SClient.Get(ctx, types.NamespacedName{Name: c.myNodeName}, &node)
+
+	return node.GetLabels(), err
```
[the remark this answers is not in the document]
Comment (haouc) [reply pkg/ipamd/ipamd.go:1232]: Yes, this is a problem. We should handle the failure on node GET.

File: pkg/ipamd/ipamd.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (achevuru) [reply pkg/ipamd/ipamd.go:650]: So, we're banking on the reconciler triggering `updateIPPoolIfRequired` every 30s?

File: pkg/ipamd/ipamd.go (hunk 15 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1368]: Yes, since this is in reconciling, we won't send event if the trunk has been attached to the node.

File: pkg/ipamd/ipamd.go (hunk 16 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1947]: This was suggested to change by go-staticcheck as `this value of pod is never used`.

Comment (haouc) [reply pkg/ipamd/ipamd.go:1947]: I can add a nil check for the pointer if that is the concern.

File: pkg/sgpp/constants.go (hunk 17 above)
Comment (haouc) [reply pkg/sgpp/constants.go:9]: We do check `c.dataStore.GetTrunkENI() == "" || !hasTrunkLabel` to send an event if a node has no trunk or has no label. We don't check label value and only check label key that means we check for `vpc.amazonaws.com/has-trunk-attached` not the boolean value.

```
if labels != nil {
		_, hasTrunkLabel = labels[trunkInterfaceLabel]
	}
```

File: pkg/ipamd/ipamd.go (hunk 16 above)
Comment (achevuru) [reply pkg/ipamd/ipamd.go:1947]: Interesting. We're using it here though - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/ipamd/ipamd.go#L1955 .

Comment (haouc) [reply pkg/ipamd/ipamd.go:1947]: I would think it was because this `pod, err = c.GetPod(podName, podNamespace)`

File: pkg/ipamd/ipamd.go (hunk 15 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1368]: We do check `c.dataStore.GetTrunkENI() == "" || !hasTrunkLabel` to send an event if a node has no trunk or has no label.

File: pkg/ipamd/ipamd.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (haouc) [reply pkg/ipamd/ipamd.go:650]: yes, Apurup. We can discuss on this offline

File: pkg/utils/eventrecorder/eventrecorder.go (hunk 18 above)
Comment (haouc) [reply pkg/utils/eventrecorder/eventrecorder.go:98]: Makes sense. I will add an error handle here.

File: pkg/ipamd/ipamd.go [hunk 22]
```diff
@@ -655,7 +658,21 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
 }
 
 func (c *IPAMContext) updateIPPoolIfRequired(ctx context.Context) {
-	c.askForTrunkENIIfNeeded(ctx)
+	if c.enablePodENI {
+		if labels, err := c.getNodeLabels(ctx); err == nil {
```
Comment (haouc) [inline pkg/ipamd/ipamd.go:662]: `getNodeLabels()` will be called every 30 seconds to ensure the correct node labels are added to nodes. This shouldn't be expensive considering it calls to cached client. We call this method only once for two labeling checks (trunk and eniConfig). One note to this change can be that we should revisit this change if we are seeing any performance impact to API server in large clusters due to this new call from every aws-node replica.

File: pkg/ipamd/ipamd.go [hunk 23]
```diff
@@ -538,37 +553,25 @@ func (c *IPAMContext) nodeInit() error {
...
-		err := c.SetNodeLabel(ctx, vpcENIConfigLabel, "")
-		if err != nil {
-			log.Errorf("Failed to delete eniConfig node label", err)
-			podENIErrInc("nodeInit")
-			return err
+
+		// if node labels is nil in rare cases, we will log the info instead of failing VPC CNI startup
+		if labels == nil {
+			// if labels is nil, ipamd logs the warning for debugging
+			log.Warn("node labels are nil during ipamd starts up, may need investigation before using the feature of security group for pods")
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:566]: nit: "ipamd starts up" ->"ipamd init"

File: pkg/ipamd/ipamd.go [hunk 24]
```diff
@@ -538,37 +553,25 @@ func (c *IPAMContext) nodeInit() error {
...
-		// Signal to VPC Resource Controller that the node has a trunk already
-		err := c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-		if err != nil {
-			log.Errorf("Failed to set node label", err)
-			podENIErrInc("nodeInit")
-			// If this fails, we probably can't talk to the API server. Let the pod restart
-			return err
+		if c.useCustomNetworking {
+			// not doing retry here, don't have to handle error since error has been logged in the called method
+			c.askForENIConfigIfNeeded(ctx, labels)
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:571]: Why do we need to proceed if labels are `nil`. Maybe we can just skip the rest of the steps?

File: pkg/ipamd/ipamd.go [hunk 25]
```diff
@@ -655,7 +658,21 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
 }
 
 func (c *IPAMContext) updateIPPoolIfRequired(ctx context.Context) {
-	c.askForTrunkENIIfNeeded(ctx)
+	if c.enablePodENI {
+		if labels, err := c.getNodeLabels(ctx); err == nil {
+			if labels == nil {
+				// if labels is nil, ipamd logs the warning for debugging
+				log.Warn("labels are nil during ipamd reconciling resources, may need attention and investigation before using the feature of security group for pods")
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:665]: Same as during nodeinit, why do we need to continue with `askForTrunkENIIfNeeded` and/or `askForENIConfigIfNeeded`

File: pkg/ipamd/ipamd.go [hunk 26]
```diff
@@ -1197,20 +1215,74 @@ func (c *IPAMContext) logPoolStats(dataStoreStats *datastore.DataStoreStats) {
 	log.Debugf("%s: %s, c.maxIPsPerENI = %d", prefix, dataStoreStats, c.maxIPsPerENI)
 }
 
-func (c *IPAMContext) askForTrunkENIIfNeeded(ctx context.Context) {
-	if c.enablePodENI && c.dataStore.GetTrunkENI() == "" {
+func (c *IPAMContext) askForTrunkENIIfNeeded(ctx context.Context, labels map[string]string) error {
+	var err error
+	hasTrunkLabel := false
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:1220]: I see here if labels are `nil`, hasTrunkLabel is false so we end up requesting a trunk ENI, how is it handled in VPC RC if trunk is attached and for whatever reason node label is missing? Do we just retry adding the label?

File: pkg/ipamd/ipamd.go [hunk 27]
```diff
@@ -1197,20 +1215,74 @@ func (c *IPAMContext) logPoolStats(dataStoreStats *datastore.DataStoreStats) {
...
+		log.Debug("VPC CNI is asking RC to initialize trunk interface")
+		err = c.eventRecorder.SendNodeEvent(corev1.EventTypeNormal, eventrecorder.EventReason, sgpp.VpcCNINodeEventActionForTrunk, sgpp.TrunkEventNote)
 		if err != nil {
-			podENIErrInc("askForTrunkENIIfNeeded")
-			log.Errorf("Failed to set node label", err)
+			log.Error("Failed sending a node event to VPC RC for enabling trunk interface")
 		}
 	}
+
+	return err
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:1244]: Btw, err is not initialized.

File: pkg/ipamd/ipamd.go [hunk 28]
```diff
@@ -1197,20 +1215,74 @@ func (c *IPAMContext) logPoolStats(dataStoreStats *datastore.DataStoreStats) {
...
+			// Signal event to VPC RC that the custom networking is enabled
+			err = c.eventRecorder.SendNodeEvent(corev1.EventTypeNormal, eventrecorder.EventReason, sgpp.VpcCNINodeEventActionForEniConfig, vpcENIConfigLabel+"="+eniConfigName)
+			if err == nil {
+				log.Infof("Send an event to notify RC custom networking is enabled. Config name should be labelled on node as %s", eniConfigName)
+			} else {
+				log.Errorf("Failed sending an event to notify RC custom networking is enabled. Error is %s", err.Error())
+			}
+		}
+	}
+	return err
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:1285]: Same here, initialize `err`.

File: pkg/ipamd/ipamd.go [hunk 29]
```diff
@@ -1197,20 +1215,74 @@ func (c *IPAMContext) logPoolStats(dataStoreStats *datastore.DataStoreStats) {
...
+	} else {
+		return node.GetLabels(), err
+	}
+}
+
+func (c *IPAMContext) askForENIConfigIfNeeded(ctx context.Context, labels map[string]string) error {
+	var err error
+	hasLabelled := false
+
+	if err != nil {
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:1260]: Why is this here? Missed during merge or cleanup?

File: pkg/ipamd/ipamd.go [hunk 30]
```diff
@@ -538,37 +553,25 @@ func (c *IPAMContext) nodeInit() error {
...
-			return err
+		if c.useCustomNetworking {
+			// not doing retry here, don't have to handle error since error has been logged in the called method
+			c.askForENIConfigIfNeeded(ctx, labels)
 		}
-	} else {
-		// Check if we want to ask for one
-		c.askForTrunkENIIfNeeded(ctx)
+		// we want to ask VPC RC once during node init to cover special cases
+		c.askForTrunkENIIfNeeded(ctx, labels)
```
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:574]: Do we skip adding events for unsupported instances? I didn't see a check or I might have missed during code walk thru.

Comment (haouc) [reply pkg/ipamd/ipamd.go:574]: We won't. VPC resource controller will check if the instance is supported. Since the instance type is checked from a static file in resource controller and so far we don't have a API to check during runtime, VPC CNI can't check and send event.

File: pkg/ipamd/ipamd.go (hunk 29 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1260]: oh, this should have been removed after I combined the getNodeLabels into caller instead of making two calls in trunk and eniconfig methods. I will remove this. Thanks.

File: pkg/ipamd/ipamd.go (hunk 28 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1285]: I think we don't have to initialize an error variable and error is nil when we declare it as an error.

File: pkg/ipamd/ipamd.go (hunk 27 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1244]: I think we don't have to init error and error is nil when we declare it as an error?

File: pkg/ipamd/ipamd.go (hunk 26 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:1220]: Yes, I am adding a check in resource controller to handle this. I don't think we handled the nil map in labels in current codes. I am thinking to either retry in resource controller in case API server error, or add a map to pod as labels and add the label.

File: pkg/ipamd/ipamd.go (hunk 23 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:566]: makes sense, will update.

File: pkg/ipamd/ipamd.go (hunk 27 above)
Comment (jayanthvn) [reply pkg/ipamd/ipamd.go:1244]: Yes defaulted to `nil`

**Review by jayanthvn — COMMENTED** [review]
Also once we merge this, we need to add few more integration tests to name a few.. CNI's behavior when label is not present, restart scenario, node not supported, vpc rc failing to add label and so on.

**Comment by haouc** [comment]
> 

Makes sense. To adding the new tests will have to wait for the control plane side update to enable the new resource controller. Thanks.

File: pkg/ipamd/ipamd.go (hunk 30 above)
Comment (haouc) [reply pkg/ipamd/ipamd.go:574]: Yes. I will update README and also check if that can be feasible to use a dummy interface on unsupported instance types.

**Comment by haouc** [comment]
I added not-supported label value check to help CNI stopping sending events from unsupported nodes. Please check the VPC resource controller PR as reference 
https://github.com/aws/amazon-vpc-resource-controller-k8s/pull/164

File: pkg/ipamd/ipamd.go [hunk 31]
```diff
@@ -134,6 +134,8 @@ const (
 
 	// trunkInterfaceLabel is used by CNI to check if correct label was added by the VPC resource controller
 	trunkInterfaceLabel = "vpc.amazonaws.com/has-trunk-attached"
+	// VPC resource controller label nodes as not-supported if the EC2 instance is not supported yet
```
Comment (achevuru) [inline pkg/ipamd/ipamd.go:137]: nit: `if the EC2 instance is not supported yet` --> `if the EC2 instance doesn't support Branch ENI feature`

Comment (haouc) [reply pkg/ipamd/ipamd.go]: Makes sense. Updated the comment.

**Review by achevuru — APPROVED** [review]

**Review by jayanthvn — APPROVED** [review]
lgtm
